openverse_catalog/dags/data_refresh/dag_factory.py

```diff
@@ -0,0 +1,197 @@
+from common.constants import DAG_DEFAULT_ARGS, XCOM_PULL_TEMPLATE
+from common.sensors.single_run_external_dags_sensor import SingleRunExternalDAGsSensor
+from data_refresh.data_refresh_types import DATA_REFRESH_CONFIGS, DataRefresh
+from requests import Response
+
+
+logger = logging.getLogger(__name__)
+
+
+DATA_REFRESH_POOL = "data_refresh"
+
+
+def response_filter_data_refresh(response: Response) -> str:
+    """
+    Filter for the `trigger_data_refresh` task, used to grab the endpoint needed
+    to poll for the status of the triggered data refresh. This information will
+    then be available via XCom in the downstream tasks.
+    """
+    status_check_url = response.json()["status_check"]
+    return urlparse(status_check_url).path
+
+
+# Response check to the `wait_for_completion` Sensor. Processes the response to
+# determine whether the task can complete.
+def response_check_wait_for_completion(response: Response) -> bool:
+    """
+    Response check to the `wait_for_completion` Sensor. Processes the response to
+    determine whether the task can complete.
+    """
+    data = response.json()
+
+    if data["active"]:
+        # The data refresh is still running. Poll again later.
+        return False
+
+    if data["error"]:
+        raise AirflowException("Error triggering data refresh.")
+
+    logger.info(
+        f"Data refresh done with {data['percent_completed']}% \
+        completed."
+    )
+    return True
+
+
+def create_data_refresh_dag(data_refresh: DataRefresh, external_dag_ids: Sequence[str]):
+    """
+    This factory method instantiates a DAG that will run the data refresh for
+    the given `media_type`.
+
+    A data refresh runs for a given media type in the API DB. It imports the
+    data for that type from the upstream DB in the Catalog, reindexes the data,
+    and updates and reindex Elasticsearch.
+
+    A data refresh can only be performed for one media type at a time, so the DAG
+    must also use a Sensor to make sure that no two data refresh tasks run
+    concurrently.
+
+    It is intended that the data_refresh tasks, or at least the initial
+    `wait_for_data_refresh` tasks, should be run in a custom pool with 1 worker
+    slot. This enforces that no two `wait_for_data_refresh` tasks can start
+    concurrently and enter a race condition.
+
+    Required Arguments:
+
+    data_refresh:     dataclass containing configuration information for the
+                      DAG
+    external_dag_ids: list of ids of the other data refresh DAGs. This DAG
+                      will not run concurrently with any dependent DAG.
+    """
+    default_args = {
+        **DAG_DEFAULT_ARGS,
+        **data_refresh.default_args,
+        "pool": DATA_REFRESH_POOL,
+    }
+    dag = DAG(
+        dag_id=data_refresh.dag_id,
+        default_args=default_args,
+        start_date=data_refresh.start_date,
+        schedule_interval=data_refresh.schedule_interval,
```

Comment:
One thing that came up while I was testing - we'll want to add `max_active_runs=1` here to ensure that only one DAG run per type is running at any given time.

Reply:
Good idea!